frappe/core/doctype/file/file.js

```diff
@@ -82,7 +82,14 @@ frappe.ui.form.on("File", {
 		if (frm.doc.file_name) {
 			file_url = file_url.replace(/#/g, "%23");
 		}
-		window.open(file_url);
+
+		// create temporary link element to simulate a download click
+		var link = document.createElement("a");
+		link.href = file_url;
+		link.download = frm.doc.file_name;
+		link.style.display = "none";
+		link.click();
+		link.remove();
```

Comment:
This might not work for all Files: as per [the MDN article about \<a\>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#download:~:text=download%20only%20works%20for%20same%2Dorigin%20URLs%2C%20or%20the%20blob%3A%20and%20data%3A%20schemes.), the `download` attribute only works for same-origin URLs (or the blob: and data: schemes). This means that I can only download files hosted on my Frappe instance.

If I host files on something like S3, or someone else's site, the download could fail.

---


Note that on some browsers/OSes, _it is believed that_ the link cannot be "click"ed if it's not in the DOM:

* `frappe.tools.downloadify`
https://github.com/frappe/frappe/blob/75412ba8728a50adff254687a56d2d9d3cec07c9/frappe/public/js/frappe/utils/tools.js#L32-L35
* Print Designer
https://github.com/frappe/print_designer/blob/29a070720395792b4ce99d9f83751e8ed94c6a46/print_designer/print_designer/client_scripts/print.js#L428

Reply:
>This might not work for all Files: as per [the MDN article about <a>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#download:~:text=download%20only%20works%20for%20same%2Dorigin%20URLs%2C%20or%20the%20blob%3A%20and%20data%3A%20schemes.), the download attribute only works for same-origin URLs (or the blob: and data: schemes). This means that I can only download files hosted on my Frappe instance.

@cogk chrome and firefox seem to auto resolve downloads to redirects if the file_urls are absolute paths where the file is hosted. Ig it's fine if it's only available self hosted files assuming more people use localfs than files via some other cdn or s3? Saves me a couple of clicks at least.

> Note that on some browsers/OSes, it is believed that the link cannot be "click"ed if it's not in the DOM

Resolved this https://github.com/frappe/frappe/pull/26544